pkg/controllers/provisioningv2/rke2/rkecluster/controller.go

```diff
@@ -37,38 +38,44 @@ func Register(ctx context.Context, clients *wrangler.Context) {
 	}, clients.RKE.RKECluster(), clients.RKE.RKEControlPlane())
 }
 
-func (h *handler) UpdateSpec(_ string, cluster *v1.RKECluster) (*v1.RKECluster, error) {
+func (h *handler) OnChange(_ string, cluster *v1.RKECluster) (*v1.RKECluster, error) {
 	if cluster == nil {
 		return nil, nil
 	}
 
-	if cluster.Spec.ControlPlaneEndpoint == nil {
+	capiCluster, err := rke2.GetOwnerCAPICluster(cluster, h.capiClusterCache)
+	if apierrors.IsNotFound(err) {
+		logrus.Debugf("[rkecluster] %s/%s: waiting: CAPI cluster does not exist", cluster.Namespace, cluster.Name)
+		h.rkeCluster.EnqueueAfter(cluster.Namespace, cluster.Name, 10*time.Second)
+		return cluster, generic.ErrSkip
+	}
+	if err != nil {
+		logrus.Errorf("[rkecluster] %s/%s: error getting CAPI cluster %v", cluster.Namespace, cluster.Name, err)
+		return cluster, err
+	}
+
+	if capiannotations.IsPaused(capiCluster, cluster) {
+		logrus.Infof("[rkecluster] %s/%s: waiting: CAPI cluster or RKECluster is paused", cluster.Namespace, cluster.Name)
+		h.rkeCluster.EnqueueAfter(cluster.Namespace, cluster.Name, 10*time.Second)
+		return cluster, generic.ErrSkip
+	}
+
+	if cluster.Spec.ControlPlaneEndpoint == nil || !cluster.Spec.ControlPlaneEndpoint.IsValid() {
 		cluster := cluster.DeepCopy()
-		cluster.Spec.ControlPlaneEndpoint = &v1.Endpoint{
+		cluster.Spec.ControlPlaneEndpoint = &capi.APIEndpoint{
 			Host: "localhost",
 			Port: 6443,
 		}
-		return h.clusterClient.Update(cluster)
+		logrus.Debugf("[rkecluster] %s/%s: setting controlplane endpoint", cluster.Namespace, cluster.Name)
+		return h.rkeCluster.Update(cluster)
 	}
 
-	return cluster, nil
-}
-
-func (h *handler) OnChange(rkeCluster *v1.RKECluster, status v1.RKEClusterStatus) (v1.RKEClusterStatus, error) {
-	conditionToUpdate := rke2.Ready
-	if !rkeCluster.DeletionTimestamp.IsZero() {
-		conditionToUpdate = rke2.Removed
-	}
-	cp, err := h.rkeControlPlanes.Get(rkeCluster.Namespace, rkeCluster.Name)
-	if err == nil {
-		conditionToUpdate.SetStatus(&status, conditionToUpdate.GetStatus(cp))
-		conditionToUpdate.Reason(&status, conditionToUpdate.GetReason(cp))
-		conditionToUpdate.Message(&status, conditionToUpdate.GetMessage(cp))
-	} else if !apierrors.IsNotFound(err) {
-		return status, err
+	if cluster.Status.Ready != true {
```

Comment:
Do we not want to do a 1-2 punch for setting controlplane endpoint and then marking the cluster as ready? Although I guess I don't know if we can have a ready cluster when the controlplane endpoint isn't set...?

Reply:
I considered that, but I chose against it for the following reason:
Our `rkecluster` object is effectively useless, solely existing to satisfy CAPI requirements. Maybe this is future proofing too much, but keeping them separate to me clarifies the intent that they are unrelated, distinct conditions that both have to be set. I'm not opposed to changing it to 1 update, as according to the upstream docs:
> The Cluster controller bubbles up spec.controlPlaneEndpoint and status.ready into status.infrastructureReady from the infrastructureCluster.

Reply:
I also had to do it this way because you can't use `Update` to update status at the same time.